### PR TITLE
update kotlinx-coroutines

### LINKED ISSUE #2911: Kotest 5.2.2 not compatible with Coroutines 1.6.1
I'm receiving the following exception when trying to run Kotest tests with the latest version 1.6.1 of Coroutines:

```
Receiver class io.kotest.assertions.ErrorCollectorContextElement does not define or inherit an implementation of the resolved method 'abstract kotlinx.coroutines.CopyableThreadContextElement copyForChild()' of interface kotlinx.coroutines.CopyableThreadContextElement.
java.lang.AbstractMethodError: Receiver class io.kotest.assertions.ErrorCollectorContextElement does not define or inherit an implementation of the resolved method 'abstract kotlinx.coroutines.CopyableThreadContextElement copyForChild()' of interface kotlinx.coroutines.CopyableThreadContextElement.
	at kotlinx.coroutines.CoroutineContextKt$foldCopies$1.invoke(CoroutineContext.kt:83)
	at kotlinx.coroutines.CoroutineContextKt$foldCopies$1.invoke(CoroutineContext.kt:80)
	at kotlin.coroutines.CoroutineContext$Element$DefaultImpls.fold(CoroutineContext.kt:70)
	at kotlinx.coroutines.ThreadContextElement$DefaultImpls.fold(ThreadContextElement.kt:52)
	at kotlinx.coroutines.CopyableThreadContextElement$DefaultImpls.fold(ThreadContextElement.kt:139)
	at io.kotest.assertions.ErrorCollectorContextElement.fold(ErrorCollector.kt:39)
	at kotlinx.coroutines.CoroutineContextKt.foldCopies(CoroutineContext.kt:80)
	at kotlinx.coroutines.CoroutineContextKt.newCoroutineContext(CoroutineContext.kt:36)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:152)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at io.kotest.engine.test.interceptors.CoroutineErrorCollectorInterceptor.intercept(CoroutineErrorCollectorInterceptor.kt:27)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:83)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.interceptors.CoroutineDispatcherFactoryInterceptor$intercept$4.invokeSuspend(coroutineDispatcherFactoryInterceptor.kt:57)
	at io.kotest.engine.test.interceptors.CoroutineDispatcherFactoryInterceptor$intercept$4.invoke(coroutineDispatcherFactoryInterceptor.kt)
	at io.kotest.engine.test.interceptors.CoroutineDispatcherFactoryInterceptor$intercept$4.invoke(coroutineDispatcherFactoryInterceptor.kt)
	at io.kotest.engine.concurrency.FixedThreadCoroutineDispatcherFactory$withDispatcher$4.invokeSuspend(FixedThreadCoroutineDispatcherFactory.kt:53)
	(Coroutine boundary)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:83)
	at io.kotest.engine.test.interceptors.SupervisorScopeInterceptor$intercept$2.invokeSuspend(SupervisorScopeInterceptor.kt:23)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:83)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:83)
	at io.kotest.engine.test.interceptors.TestFinishedInterceptor.intercept(TestFinishedInterceptor.kt:21)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:83)
	at io.kotest.engine.spec.runners.SingleInstanceSpecRunner.runTest(SingleInstanceSpecRunner.kt:112)
	at io.kotest.engine.spec.runners.SingleInstanceSpecRunner$execute$interceptAndRun$2$2.invokeSuspend(SingleInstanceSpecRunner.kt:51)
	at io.kotest.engine.test.scheduler.SequentialTestScheduler$schedule$3$1$1.invokeSuspend(SequentialTestScheduler.kt:20)
	at io.kotest.engine.test.scheduler.SequentialTestScheduler.schedule(SequentialTestScheduler.kt:18)
	at io.kotest.engine.spec.runners.SingleInstanceSpecRunner.execute_gIAlu_s$interceptAndRun(SingleInstanceSpecRunner.kt:49)
	at io.kotest.engine.spec.runners.SingleInstanceSpecRunner$execute$3.invokeSuspend(SingleInstanceSpecRunner.kt:58)
	at io.kotest.engine.spec.runners.SingleInstanceSpecRunner.execute-gIAlu-s(SingleInstanceSpecRunner.kt:56)
	at io.kotest.engine.spec.JvmSpecExecutorDelegate.execute(createSpecExecutorDelegate.kt:71)
	at io.kotest.engine.spec.SpecExecutor$specInterceptors$initial$1.invokeSuspend(SpecExecutor.kt:107)
	at io.kotest.engine.spec.interceptor.ConfigurationInContextInterceptor$intercept$2.invokeSuspend(ConfigurationInContextInterceptor.kt:23)
	at io.kotest.engine.spec.interceptor.ConfigurationInContextInterceptor.intercept-0E7RQCE(ConfigurationInContextInterceptor.kt:22)
	at io.kotest.engine.spec.SpecExecutor$specInterceptors$3$1.invokeSuspend(SpecExecutor.kt:116)
	at io.kotest.engine.spec.interceptor.SpecExtensionInterceptor$intercept$2.invokeSuspend(SpecExtensionInterceptor.kt:21)
	at io.kotest.engine.spec.SpecExtensions$intercept$initial$1.invokeSuspend(SpecExtensions.kt:143)
	at io.kotest.engine.spec.SpecExtensions.intercept(SpecExtensions.kt:154)
	at io.kotest.engine.spec.interceptor.SpecExtensionInterceptor.intercept-0E7RQCE(SpecExtensionInterceptor.kt:21)
	at io.kotest.engine.spec.SpecExecutor$specInterceptors$3$1.invokeSuspend(SpecExecutor.kt:116)
	at io.kotest.engine.spec.interceptor.ProjectContextInterceptor$intercept$2.invokeSuspend(ProjectContextInterceptor.kt:21)
	at io.kotest.engine.spec.interceptor.ProjectContextInterceptor.intercept-0E7RQCE(ProjectContextInterceptor.kt:20)
	at io.kotest.engine.spec.SpecExecutor$specInterceptors$3$1.invokeSuspend(SpecExecutor.kt:116)
	at io.kotest.engine.spec.SpecExecutor.specInterceptors-gIAlu-s(SpecExecutor.kt:117)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$innerExecute$1.invokeSuspend(SpecExecutor.kt:85)
	at io.kotest.engine.spec.interceptor.FinalizeSpecInterceptor.intercept-0E7RQCE(FinalizeSpecInterceptor.kt:19)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.PrepareSpecInterceptor.intercept-0E7RQCE(PrepareSpecInterceptor.kt:20)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.ApplyExtensionsInterceptor.intercept-0E7RQCE(ApplyExtensionsInterceptor.kt:35)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.SpecFinishedInterceptor.intercept-0E7RQCE(SpecFinishedInterceptor.kt:19)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.SpecStartedInterceptor.intercept-0E7RQCE(SpecStartedInterceptor.kt:19)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.SpecRefExtensionInterceptor$intercept$inner$1.invokeSuspend(SpecRefExtensionInterceptor.kt:24)
	at io.kotest.engine.spec.interceptor.SpecRefExtensionInterceptor.intercept-0E7RQCE(SpecRefExtensionInterceptor.kt:27)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.RequiresTagSpecInterceptor.intercept-0E7RQCE(RequiresTagSpecInterceptor.kt:35)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.TagsExcludedSpecInterceptor.intercept-0E7RQCE(TagsExcludedDiscoveryExtension.kt:32)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.SystemPropertySpecFilterInterceptor.intercept-0E7RQCE(SystemPropertySpecFilterInterceptor.kt:49)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.SpecFilterInterceptor.intercept-0E7RQCE(SpecFilterInterceptor.kt:37)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.IgnoredSpecInterceptor.intercept-0E7RQCE(IgnoredSpecInterceptor.kt:42)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.spec.interceptor.EnabledIfSpecInterceptor.intercept-0E7RQCE(EnabledIfSpecInterceptor.kt:40)
	at io.kotest.engine.spec.SpecExecutor$referenceInterceptors$3$1.invokeSuspend(SpecExecutor.kt:90)
	at io.kotest.engine.ConcurrentTestSuiteScheduler$schedule$8$1$2.invokeSuspend(ConcurrentTestSuiteScheduler.kt:71)
Caused by: java.lang.AbstractMethodError: Receiver class io.kotest.assertions.ErrorCollectorContextElement does not define or inherit an implementation of the resolved method 'abstract kotlinx.coroutines.CopyableThreadContextElement copyForChild()' of interface kotlinx.coroutines.CopyableThreadContextElement.
	at kotlinx.coroutines.CoroutineContextKt$foldCopies$1.invoke(CoroutineContext.kt:83)
	at kotlinx.coroutines.CoroutineContextKt$foldCopies$1.invoke(CoroutineContext.kt:80)
	at kotlin.coroutines.CoroutineContext$Element$DefaultImpls.fold(CoroutineContext.kt:70)
	at kotlinx.coroutines.ThreadContextElement$DefaultImpls.fold(ThreadContextElement.kt:52)
	at kotlinx.coroutines.CopyableThreadContextElement$DefaultImpls.fold(ThreadContextElement.kt:139)
	at io.kotest.assertions.ErrorCollectorContextElement.fold(ErrorCollector.kt:39)
	at kotlinx.coroutines.CoroutineContextKt.foldCopies(CoroutineContext.kt:80)
	at kotlinx.coroutines.CoroutineContextKt.newCoroutineContext(CoroutineContext.kt:36)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:152)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at io.kotest.engine.test.interceptors.CoroutineErrorCollectorInterceptor.intercept(CoroutineErrorCollectorInterceptor.kt:27)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:83)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.interceptors.CoroutineDispatcherFactoryInterceptor$intercept$4.invokeSuspend(coroutineDispatcherFactoryInterceptor.kt:57)
	at io.kotest.engine.test.interceptors.CoroutineDispatcherFactoryInterceptor$intercept$4.invoke(coroutineDispatcherFactoryInterceptor.kt)
	at io.kotest.engine.test.interceptors.CoroutineDispatcherFactoryInterceptor$intercept$4.invoke(coroutineDispatcherFactoryInterceptor.kt)
	at io.kotest.engine.concurrency.FixedThreadCoroutineDispatcherFactory$withDispatcher$4.invokeSuspend(FixedThreadCoroutineDispatcherFactory.kt:53)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### PR DESCRIPTION
**Pull request by i-walker**
resolves #2911 
~~the merge function was created due to https://github.com/Kotlin/kotlinx.coroutines/pull/3227/files in 1.6.1~~

### CODE REVIEW

**Comment by OliverO2** [comment]
`ErrorCollectorContextElement`, being a private class, will never be explicitly passed to a newly created child coroutine. So there is no need for a `merge()` function in `ErrorCollector` and merging `mergeForChild()` could be implemented like this:
```kotlin
override fun mergeForChild(overwritingElement: CoroutineContext.Element): CoroutineContext = copyForChild()
```

This would be in line with the example comment in `ThreadContextElement.kt`:
https://github.com/Kotlin/kotlinx.coroutines/blob/a5dd74b2b325113586768b8b61af6e2833a139c2/kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt#L127-L128

File: kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/ErrorCollector.kt [hunk 1]
```diff
@@ -45,6 +45,8 @@ interface ErrorCollector {
     * That is all the clues nested to this point.
     */
    fun clueContext(): List<Clue>
+
+   fun merge(errorCollector: ErrorCollector): ErrorCollector
```
Comment (OliverO2) [inline kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/ErrorCollector.kt]: Seems this is not required, see separate comment above.

File: kotest-assertions/kotest-assertions-shared/src/jvmMain/kotlin/io/kotest/assertions/ErrorCollector.kt [hunk 2]
```diff
@@ -52,9 +55,13 @@ private class ErrorCollectorContextElement(private val coroutineLocalErrorCollec
...
-
    companion object Key : CoroutineContext.Key<ErrorCollectorContextElement>
+
+   override fun copyForChild(): CopyableThreadContextElement<CoroutineLocalErrorCollector> =
+      ErrorCollectorContextElement(threadLocalErrorCollector.get().copy())
+
+   override fun mergeForChild(overwritingElement: CoroutineContext.Element): CoroutineContext {
+      overwritingElement as ErrorCollectorContextElement
+      return ErrorCollectorContextElement(coroutineLocalErrorCollector.merge(overwritingElement.coroutineLocalErrorCollector))
+   }
```
Comment (OliverO2) [inline kotest-assertions/kotest-assertions-shared/src/jvmMain/kotlin/io/kotest/assertions/ErrorCollector.kt]: Could be simplified to:
```kotlin
override fun mergeForChild(overwritingElement: CoroutineContext.Element): CoroutineContext = copyForChild()
```

**Review by OliverO2 — CHANGES_REQUESTED** [review]
Version 1.6.1 of kotlinx-coroutines adds a `@DelicateCoroutinesApi` annotation to `CopyableThreadContextElement`, so `ErrorCollectorContextElement` should be annotated accordingly:
```kotlin
@OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
```

**Comment by i-walker** [comment]
Sure, I'll add those changes :D

**Comment by OliverO2** [comment]
Almost: Could you add the additional annotation as requested in https://github.com/kotest/kotest/pull/2912#issuecomment-1088981841? (I apologize for my review comment on this appearing in the wrong place.)

**Review by OliverO2 — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**